File: mixin-base/src/main/python/mixin/rmo/base_runmodel.py
#
# Copyright (c) 2020-2023 Datamixin.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
from typing import Dict
from mixin.rmo.runmodel import Runmodel
from mixin.model.eobject import EObject
from mixin.rmo.outset import OutsetFactory
from mixin.rmo.indication import Indication
from mixin.rmo.supervisor import Supervisor
from mixin.rmo.modification import Modification
from mixin.rmo.root_regulator import RootRegulator
from mixin.rmo.eobject_modifier import EObjectModifier
from mixin.rmo.regulator import Regulator, RegulatorFactory


class BaseRunmodel(Runmodel):
    def __init__(self):
        super().__init__()
        self._regulatorFactory: RegulatorFactory = None
        self._outsetFactory: OutsetFactory = None
        self._capabilities: Dict[any, any] = {}
        self._rootRegulator: RootRegulator = RootRegulator()
        self._rootRegulator.setRunmodel(self)

    def setRegulatorFactory(self, factory: RegulatorFactory):
        self._regulatorFactory = factory

    def setOutsetFactory(self, factory: OutsetFactory):
        self._outsetFactory = factory

    def getModel(self) -> any:
        contents: Regulator = self._rootRegulator.getContents()
        return contents.getModel()

    def modify(self, modification: Modification):
        regulator: Regulator = self._rootRegulator.getContents()
        model: any = regulator.getModel()
        if isinstance(model, EObject):
            modifier: EObjectModifier = EObjectModifier(modification)
            modifier.modify(model)

    def getRegulatorFactory(self) -> RegulatorFactory:
        return self._regulatorFactory

    def getOutsetFactory(self) -> OutsetFactory:
        return self._outsetFactory

    def registerCapability(self, capabilityClass: any, service: any):
        self._capabilities[capabilityClass] = service

    def getCapability(self, capabilityClass: any) -> any:

        capability: any = self._capabilities.get(capabilityClass)

        # Cari menggunakan class sebagai interface
        if capability is None:
            for keyClass in self._capabilities.keys():
                value: any = self._capabilities[keyClass]
                valueClass: any = type(value)
                if issubclass(keyClass, capabilityClass):
                    capability = value
                    break
                if issubclass(valueClass, capabilityClass):
                    capability = value
                    break

        return capability

    def getRootRegulator(self) -> RootRegulator:
        return self._rootRegulator

    def getSupervisor(self) -> Supervisor:
        return self._rootRegulator.getContents()

    def postIndication(self, indication: Indication):
        pass

    def postRectification(self, rectification: Modification):
        pass

File: mixin-base/src/main/java/com/andia/mixin/rmo/BaseRunmodel.java
/*
 * Copyright (c) 2020-2023 Datamixin.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. */
package com.andia.mixin.rmo;

import java.util.HashMap;
import java.util.Map;

import com.andia.mixin.model.EObject;

public abstract class BaseRunmodel implements Runmodel {

	private RegulatorFactory regulatorFactory;
	private OutsetFactory outsetFactory;
	private RootRegulator rootRegulator = new RootRegulator();
	private Map<Class<?>, Object> capabilities = new HashMap<>();

	public BaseRunmodel() {
		rootRegulator.setRunmodel(this);
	}

	protected void setRegulatorFactory(RegulatorFactory factory) {
		this.regulatorFactory = factory;
	}

	protected void setOutsetFactory(OutsetFactory factory) {
		this.outsetFactory = factory;
	}

	@Override
	public Object getModel() {
		Regulator contents = rootRegulator.getContents();
		return contents.getModel();
	}

	public void modify(Modification modification) {
		Regulator regulator = rootRegulator.getContents();
		Object model = regulator.getModel();
		if (model instanceof EObject) {
			EObject eObject = (EObject) model;
			EObjectModifier modifier = new EObjectModifier(modification);
			modifier.modify(eObject);
		}
	}

	@Override
	public RegulatorFactory getRegulatorFactory() {
		return this.regulatorFactory;
	}

	@Override
	public OutsetFactory getOutsetFactory() {
		return outsetFactory;
	}

	@Override
	public void registerCapability(Class<?> capabilityClass, Object service) {
		this.capabilities.put(capabilityClass, service);
	}

	@Override
	public <C> C getCapability(Class<? extends C> capabilityClass) {

		Object capability = capabilities.get(capabilityClass);

		// Cari menggunakan class sebagai interface
		if (capability == null) {
			for (Class<?> keyClass : capabilities.keySet()) {
				Object value = capabilities.get(keyClass);
				Class<?> valueClass = value.getClass();
				if (capabilityClass.isAssignableFrom(keyClass)) {
					capability = value;
					break;
				}
				if (capabilityClass.isAssignableFrom(valueClass)) {
					capability = value;
					break;
				}
			}
		}

		return capabilityClass.cast(capability);
	}

	public RootRegulator getRootRegulator() {
		return this.rootRegulator;
	}

	@Override
	public Supervisor getSupervisor() {
		return this.rootRegulator.getContents();
	}

	@Override
	public void postIndication(Indication indication) {

	}

	@Override
	public void postRectification(Modification rectification) {

	}

}
